diff --git a/src/tools/gpiomon.c b/src/tools/gpiomon.c
index c13ab38..d07e7aa 100644
--- a/src/tools/gpiomon.c
+++ b/src/tools/gpiomon.c
@@ -27,10 +27,11 @@ static const struct option longopts[] = {
 	{ "rising-edge",	no_argument,		NULL,	'r' },
 	{ "falling-edge",	no_argument,		NULL,	'f' },
 	{ "format",		required_argument,	NULL,	'F' },
+	{ "script",		required_argument,	NULL,	'S' },
 	{ GETOPT_NULL_LONGOPT },
 };
 
-static const char *const shortopts = "+hvln:srfF:";
+static const char *const shortopts = "+hvln:srfF:S:";
 
 static void print_help(void)
 {
@@ -53,6 +54,7 @@ static void print_help(void)
 	printf("  %%e:  event type (0 - falling edge, 1 rising edge)\n");
 	printf("  %%s:  seconds part of the event timestamp\n");
 	printf("  %%n:  nanoseconds part of the event timestamp\n");
+	printf("  -S, --script=SCT\tspecify script to run upon completion\n");
 }
 
 struct mon_ctx {
@@ -62,6 +64,7 @@ struct mon_ctx {
 
 	bool silent;
 	char *fmt;
+	char *sct;
 
 	int sigfd;
 };
@@ -212,8 +215,10 @@ static int event_callback(int event_type, unsigned int line_offset,
 		return GPIOD_CTXLESS_EVENT_CB_RET_OK;
 	}
 
-	if (ctx->events_wanted && ctx->events_done >= ctx->events_wanted)
+	if (ctx->events_wanted && ctx->events_done >= ctx->events_wanted) {
+		system(ctx->sct);
 		return GPIOD_CTXLESS_EVENT_CB_RET_STOP;
+	}
 
 	return GPIOD_CTXLESS_EVENT_CB_RET_OK;
 }
@@ -281,6 +286,9 @@ int main(int argc, char **argv)
 		case 'F':
 			ctx.fmt = optarg;
 			break;
+		case 'S':
+			ctx.sct = optarg;
+			break;
 		case '?':
 			die("try %s --help", get_progname());
 		default:
